### PR TITLE
fix(2446): fix environment evaluation order

### PR DESCRIPTION
**Pull request by itleigns**
## Context
The order of evaluation of environment values is not consistent in sd-local. We want to fix it to the order in which they are defined in `screwdriver.yaml`.

## Objective
- We change the data structure of the environment to keep the order in which they are defined
- We evaluate environment variables that contain environment variables in `launch.go`

## References
[screwdriver-cd/screwdriver#2446](https://github.com/screwdriver-cd/screwdriver/issues/2446)

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: screwdriver/screwdriver.go [hunk 1]
```diff
@@ -41,11 +42,31 @@ type Step struct {
 	Command string `json:"command"`
 }
 
+// MapSlice is an associative array which preserves order
+type MapSlice struct {
+	Body []map[string]string
```
Comment (tanoda) [inline screwdriver/screwdriver.go:47]: it should be sequence of `struct { Key, Value string }`, because this map contains single key/value only

File: launch/launch.go [hunk 2]
```diff
@@ -102,9 +104,17 @@ func DefaultSocketPath() string {
...
 
-func mergeEnv(env, jobEnv, optionEnv EnvVar) []EnvVar {
-	for k, v := range jobEnv {
-		env[k] = v
+func mergeEnv(env EnvVar, jobEnv screwdriver.MapSlice, optionEnv EnvVar) []EnvVar {
+	for _, e := range jobEnv.Body {
+		for k, v := range e {
+			val := v
+			for fr, to := range env {
+				val = strings.Replace(val, "${"+fr+"}", to, -1)
```
Comment (wahapo) [inline launch/launch.go:112]: Environment variables can be specified not only as `${FOO}' but also as `$FOO`.

File: screwdriver/screwdriver.go (hunk 1 above)
Comment (itleigns) [reply screwdriver/screwdriver.go]: I fixed it.

File: launch/launch.go (hunk 2 above)
Comment (itleigns) [reply launch/launch.go]: I fixed it. I use `os.ExpandEnv`.

**Review by wahapo — DISMISSED** [review]

**Review by ibu1224 — APPROVED** [review]

File: screwdriver/screwdriver.go [hunk 3]
```diff
@@ -41,11 +42,39 @@ type Step struct {
...
+// PairSlice is an associative array which preserves order
+type PairSlice struct {
+	Body []struct {
+		Key   string
+		Value string
+	}
+}
+
+// UnmarshalJSON replaces JSON of a normal associative array to PairSlice
+func (MS *PairSlice) UnmarshalJSON(data []byte) error {
```
Comment (yuichi10) [inline screwdriver/screwdriver.go:54]: Why the name of reciver is `MS`?
Is it better to use `p` or `ps`?

Comment (itleigns) [reply screwdriver/screwdriver.go]: I forgot to change the name. I fixed it.

File: screwdriver/screwdriver.go [hunk 4]
```diff
@@ -41,11 +42,39 @@ type Step struct {
 	Command string `json:"command"`
 }
 
+// PairSlice is an associative array which preserves order
+type PairSlice struct {
```
Comment (yuichi10) [inline screwdriver/screwdriver.go:46]: [IMO]
It seems to be easier to understand without the Body.
How about this?
```
type Environments []struct{
    Key string
    Value string
}
```
or 
```
type Pair struct {
    Key string
    Value string
}
type Environments []Pair
```

File: launch/launch.go [hunk 5]
```diff
@@ -102,9 +102,11 @@ func DefaultSocketPath() string {
...
 }
 
-func mergeEnv(env, jobEnv, optionEnv EnvVar) []EnvVar {
-	for k, v := range jobEnv {
-		env[k] = v
+func mergeEnv(env EnvVar, jobEnv screwdriver.PairSlice, optionEnv EnvVar) []EnvVar {
+	for _, e := range jobEnv.Body {
+		val := os.ExpandEnv(e.Value)
+		env[e.Key] = val
+		os.Setenv(e.Key, val)
```
Comment (tk3fftk) [inline launch/launch.go:109]: Do we need an error handling?

File: screwdriver/screwdriver.go [hunk 6]
```diff
@@ -41,11 +42,39 @@ type Step struct {
...
+	}
+}
+
+// UnmarshalJSON replaces JSON of a normal associative array to PairSlice
+func (PS *PairSlice) UnmarshalJSON(data []byte) error {
+	var output []struct {
+		Key   string
+		Value string
+	}
+	r := regexp.MustCompile(`"(.*?)" *: *"(.*?)"`)
```
Comment (tk3fftk) [inline screwdriver/screwdriver.go:59]: `regexp.MustCompile` should be global for safe and to decrease initializing cost.

https://golang.org/pkg/regexp/#MustCompile

File: screwdriver/screwdriver.go (hunk 4 above)
Comment (tk3fftk) [reply screwdriver/screwdriver.go:46]: In addition, `Environments` looks better name.

Comment (itleigns) [reply screwdriver/screwdriver.go]: I fixed it.

File: launch/launch.go (hunk 5 above)
Comment (itleigns) [reply launch/launch.go]: I added error handling.

File: screwdriver/screwdriver.go (hunk 6 above)
Comment (itleigns) [reply screwdriver/screwdriver.go]: I fixed it.

File: screwdriver/screwdriver.go [hunk 7]
```diff
@@ -41,11 +44,31 @@ type Step struct {
...
 }
 
+// Environments is an environment slice
+type Environments []struct {
+	Key   string
+	Value string
+}
+
+// UnmarshalJSON replaces JSON of a normal associative array to Environments
+func (en *Environments) UnmarshalJSON(data []byte) error {
```
Comment (yuichi10) [inline screwdriver/screwdriver.go:54]: Who uses this function? I cannot find the place that use this function

Comment (yuichi10) [reply screwdriver/screwdriver.go:54]: Sorry now i understand that this implement interface.

**Review by yuichi10 — APPROVED** [review]
LGTM

File: screwdriver/screwdriver.go [hunk 8]
```diff
@@ -8,10 +8,13 @@ import (
 	"net/http"
 	"net/url"
 	"path"
+	"regexp"
 	"strconv"
 	"strings"
 )
 
+var regexpForUnmarshal = regexp.MustCompile(`"(.*?)" *: *"(.*?)"`)
```
Comment (tk3fftk) [inline screwdriver/screwdriver.go:16]: Do we need white space before `*` ?
(I don't know what kind of string will be come.)

Comment (itleigns) [reply screwdriver/screwdriver.go:16]: Basically, we do not need the white spaces. However we need them in test because the jsons we are using for testing contain whitespace.

**Review by tk3fftk — APPROVED** [review]
👍
